src/cryptenroll/cryptenroll-tpm2.c

```diff
@@ -160,6 +163,21 @@ int enroll_tpm2(struct crypt_device *cd,
                 r = get_pin(&pin_str, &flags);
                 if (r < 0)
                         return r;
+
+                r = crypto_random_bytes(binary_salt, sizeof(binary_salt));
+                if (r != 0)
+                        return 1;
+
+                uint8_t salted_pin[SHA256_DIGEST_SIZE] = {};
+                tpm2_util_pbkdf(pin_str, strlen(pin_str), binary_salt, sizeof(binary_salt), 1000, salted_pin);
+
+                explicit_bzero_safe(pin_str, strlen(pin_str));
```

Comment:
Adding explicit_bzero_safe and mfree() as using erase_and_freep was leading to a segfault

Reply:
note that erase_and_free() and erase_and_freep() are two different things. The difference being that the former takes a pointer to with the memory to free, the latter takes a pointer to a pointer with the memory to free. The latter exists only to make the `_cleanup_()` logic work, where you get a pointer to a pointer for memory allocations.

With other words, use `pin = erase_and_free(pin);` here, and all is good.

Reply:
> note that erase_and_free() and erase_and_freep() are two different things. The difference being that the former takes a pointer to with the memory to free, the latter takes a pointer to a pointer with the memory to free. The latter exists only to make the _cleanup_() logic work, where you get a pointer to a pointer for memory allocations.

Ahhh, thanks I knew I was missing something in the _cleanup_ callbacks, they pass a double pointer and now that trailing `p` makes sense.